File: Minimum BST Sum Subtree, 18th April.cpp
vector<int> help(Node* root,int target,int &ans){
        if(!root)return {1,0,0,INT_MAX,INT_MIN};
        vector<int> l = help(root->left,target,ans);
        vector<int> r = help(root->right,target,ans);
        if(root->data<=l[4] || root->data>=r[3])return {0,0,0,0,0};
        int sum = root->data+l[2]+r[2];
        int n = 1+l[1]+r[1];
        if(l[0]==1 && r[0]==1 && sum==target){
            ans=min(ans,n);
        }
        return {l[0]&r[0],n,sum,min(l[3],root->data),max(r[4],root->data)};
        
    }
    int minSubtreeSumBST(int target, Node *root) {
        // code here
        int ans = INT_MAX;
        vector<int> temp = help(root,target,ans);
        if(temp[0]==1 and temp[2]==target){
            ans=min(ans,temp[1]);
        }
        if(ans==INT_MAX)return -1;
        return ans;
    }

File: Minimum BST Sum Subtree, 18th April.py
#User function Template for python3

'''
class Node:
    def __init__(self,val):
        self.data=val
        self.left=None
        self.right=None
'''

class Solution:
    def check(self,node,mini,maxi):
        
        if node is None:
            return True
 
    # False if this node violates min/max constraint
        if node.data < mini or node.data > maxi:
           return False
 
    # Otherwise check the subtrees recursively
    # tightening the min or max constraint
        return (self.check(node.left, mini, node.data - 1) and
            self.check(node.right, node.data+1, maxi))
     
    def solve(self,root,ans,target,k):
        maxi=4294967296
        mini = -4294967296
        if root==None:
            return (0,0)
        left,l=self.solve(root.left,ans,target,k)
        right,r=self.solve(root.right,ans,target,k)
        
        if left+right+root.data==target and self.check(root,mini,maxi)==True :
            if self.s  > l+r+1:
                self.s=l+1+r
                
            
         
        return (left+right+root.data, l+r+1 )   
    def minSubtreeSumBST(self, target, root):
        
        self.s=10000   
        ans,s=self.solve(root,0,target,0)
        
        if self.s==10000:
            return -1
        return self.s
